<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Abstract gRNA · ARTEMIS.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/theme.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img class="docs-light-only" src="../assets/logo.png" alt="ARTEMIS.jl logo"/><img class="docs-dark-only" src="../assets/logo-dark.png" alt="ARTEMIS.jl logo"/></a><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">General</a></li><li><span class="tocitem">API</span><ul><li class="is-active"><a class="tocitem" href>Abstract gRNA</a></li><li><a class="tocitem" href="../find_potential_ot/">Find potential off-targets</a></li><li><a class="tocitem" href="../align_gRNA/">Align gRNA and off-target</a></li><li><a class="tocitem" href="../alignment_free/">Alignment-free filters for gRNAs</a></li><li><a class="tocitem" href="../find_ot/">Find all off-targets</a></li><li><a class="tocitem" href="../utils/">Utils</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>Abstract gRNA</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Abstract gRNA</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Abstract-gRNA"><a class="docs-heading-anchor" href="#Abstract-gRNA">Abstract gRNA</a><a id="Abstract-gRNA-1"></a><a class="docs-heading-anchor-permalink" href="#Abstract-gRNA" title="Permalink"></a></h1><p>To encapsulate motifs that we search for on the genome, we can use abstract definition of gRNA/off-target, which is defined in <code>Motif</code> data structure.</p><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.Motif" href="#ARTEMIS.Motif"><code>ARTEMIS.Motif</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Motif(
    alias::String, 
    fwdmotif::String, 
    fwdpam::String, 
    forward_strand::Bool = true, 
    reverse_strand::Bool = true, 
    distance::Int = 4, 
    extends5::Bool = true,
    ambig_max::Int = 5)</code></pre><p><code>Motif(alias::String)</code></p><p>Motif defines what we search on the genome, what can be identified as an off-target.</p><p><strong>Arguments</strong></p><p><code>alias</code> - alias of the motif for easier identification e.g. Cas9</p><p><code>fwdmotif</code> - Motif that indicates where is PAM inside <code>fwdpam</code>.     For example for Cas9 it is 20*N + XXX: NNNNNNNNNNNNNNNNNNNNXXX</p><p><code>fwdpam</code>   - Motif in 5&#39;-3&#39; that will be matched on the reference (without the X).              For example for Cas9 it is 20*X + NGG:              XXXXXXXXXXXXXXXXXXXXNGG</p><p><code>forward</code>  - If false will not match to the forward reference strand.</p><p><code>reverse</code>  - If false will not match to the reverse reference strand.</p><p><code>distance</code> - How many extra nucleotides are needed for a search? This              will indicate within what distance we can search for off-targets.              When we don&#39;t have those bases we use DNA_Gap.</p><p><code>extend5</code>  - Defines how off-targets will be aligned to the guides and where              extra nucleotides will be added for alignment within distance. Whether              to extend in the 5&#39; and 3&#39; direction. Cas9 is extend5 = true.</p><p><code>ambig_max</code>- How many ambiguous bases are allowed in the pattern?</p><pre><code class="nohighlight hljs">Example for Cas9 where we want to search for off-targets within distance of 4:
  alias:    Cas9
  fwdmotif: NNNNNNNNNNNNNNNNNNNNXXX
  fwdpam:   XXXXXXXXXXXXXXXXXXXXNGG
  forward:  true
  reverse:  true
  distance: 4
  extend5:  true
  ambig_max:5 </code></pre><p>Alignments will be performed from opposite to the extension direction (which is defined by extend5).</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; Motif(&quot;Cas9&quot;)
Motif(&quot;Cas9&quot;, NNNNNNNNNNNNNNNNNNNNNGG, CCNNNNNNNNNNNNNNNNNNNNN, 21:23, 1:3, 3, true, 0)

julia&gt; Motif(&quot;test name&quot;, &quot;NNNNNNNNNNNNNNNNNNNNXXX&quot;, &quot;XXXXXXXXXXXXXXXXXXXXNGG&quot;, true, true, 3, true, 5)
Motif(&quot;test name&quot;, NNNNNNNNNNNNNNNNNNNNNGG, CCNNNNNNNNNNNNNNNNNNNNN, 21:23, 1:3, 3, true, 5)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.length_noPAM" href="#ARTEMIS.length_noPAM"><code>ARTEMIS.length_noPAM</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>length_noPAM(motif::Motif)</code></p><p>Calculate what is the length of the motif, without extension, and without PAM. Effectively, size of the gRNA.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; length_noPAM(Motif(&quot;Cas9&quot;))
20</code></pre></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.setdist" href="#ARTEMIS.setdist"><code>ARTEMIS.setdist</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>setdist(motif::Motif, distance::Int)</code></p><p>Set the distance (maximum value of allowed mismatches, deletion, insertions)  that are allowed during alignment.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; setdist(Motif(&quot;Cas9&quot;), 15)
Motif(&quot;Cas9&quot;, NNNNNNNNNNNNNNNNNNNNNGG, CCNNNNNNNNNNNNNNNNNNNNN, 21:23, 1:3, 15, true, 0)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.setambig" href="#ARTEMIS.setambig"><code>ARTEMIS.setambig</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>setambig(motif::Motif, ambig::Int)</code></p><p>Set the ambiguity (how many ambiguous bases are allowed, not counting PAM, not counting extension) level for <code>motif</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; setambig(Motif(&quot;Cas9&quot;), 15)
Motif(&quot;Cas9&quot;, NNNNNNNNNNNNNNNNNNNNNGG, CCNNNNNNNNNNNNNNNNNNNNN, 21:23, 1:3, 3, true, 15)</code></pre></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« General</a><a class="docs-footer-nextpage" href="../find_potential_ot/">Find potential off-targets »</a><div class="flexbox-break"></div><p class="footer-message">Made with ♥ by Kornel Labun.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Wednesday 2 November 2022 12:48">Wednesday 2 November 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
