<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · CRISPRofftargetHunter.jl</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>CRISPRofftargetHunter.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Citation"><span>Citation</span></a></li><li><a class="tocitem" href="#LICENSE"><span>LICENSE</span></a></li><li><a class="tocitem" href="#Self-contained-build"><span>Self-contained build</span></a></li><li><a class="tocitem" href="#Main-API"><span>Main API</span></a></li><li><a class="tocitem" href="#Example"><span>Example</span></a></li><li><a class="tocitem" href="#Public-Interface"><span>Public Interface</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="CRISPRofftargetHunter"><a class="docs-heading-anchor" href="#CRISPRofftargetHunter">CRISPRofftargetHunter</a><a id="CRISPRofftargetHunter-1"></a><a class="docs-heading-anchor-permalink" href="#CRISPRofftargetHunter" title="Permalink"></a></h1><p>Software designed for efficient, precise and fast identification of all off-targets for given CRISPR guideRNA&#39;s. It leverages couple of alghoritms designed specifically for this purpose.</p><p>CRISPRofftargetHunter is <strong>extensively tested</strong>:</p><ul><li>unit tests for each function</li><li>friction tests where three different implementations of the same functionality must report the same results</li><li>end-to-end tests where we run whole pipeline on specially designed sample genome and compare results with CRISPRitz software</li></ul><p>CRISPRofftargetHunter is designed specifically for CRISPR alignments of guideRNA&#39;s, <strong>allowing for deletions, insretions and mismatches</strong>. Implemented alghoritms allow you to find off-targets within <strong>distance as large as you want</strong>!</p><p>CRISPRofftargetHunter has an alghoritm (see: <code>build_sketchDB</code> and <code>search_sketchDB</code>) designed for <strong>super fast estimation of number of off-targets</strong> in the genome. These estimations can never report counts less than reality, but can only over-estimate! This can be used to quickly design libraries of guideRNA&#39;s for the entire genomes, as promising guides can be quickly sorted using CRISPRofftargetHunter.</p><p>CRISPRofftargetHunter has <strong>support for multiple-cores</strong>, we use standard julia configuration for that.</p><h2 id="Citation"><a class="docs-heading-anchor" href="#Citation">Citation</a><a id="Citation-1"></a><a class="docs-heading-anchor-permalink" href="#Citation" title="Permalink"></a></h2><p>TODO</p><h2 id="LICENSE"><a class="docs-heading-anchor" href="#LICENSE">LICENSE</a><a id="LICENSE-1"></a><a class="docs-heading-anchor-permalink" href="#LICENSE" title="Permalink"></a></h2><p>Copyright (C) 2021  Kornel Labun</p><p>This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.</p><p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.</p><p>You should have received a copy of the GNU Affero General Public License along with this program.  If not, see <a href="https://www.gnu.org/licenses/">https://www.gnu.org/licenses/</a>.</p><h2 id="Self-contained-build"><a class="docs-heading-anchor" href="#Self-contained-build">Self-contained build</a><a id="Self-contained-build-1"></a><a class="docs-heading-anchor-permalink" href="#Self-contained-build" title="Permalink"></a></h2><p>It is possible to build CRISPRofftargetHunter into standalone application. This can be achieved by running <code>./build_standalone.sh</code> script from the main directory. Script will produce binary in a new folder outside the main directory. Then you can run from inside that folder <code>./bin/CRISPRofftargetHunter --help</code> or <code>./bin/CRISPRofftargetHunter build --help</code>.</p><h2 id="Main-API"><a class="docs-heading-anchor" href="#Main-API">Main API</a><a id="Main-API-1"></a><a class="docs-heading-anchor-permalink" href="#Main-API" title="Permalink"></a></h2><p>Run CRISPRofftargetHunter as an application. From the directory of the package run:</p><pre><code class="language-bash">julia --threads 4 --project=&quot;.&quot; CRISPRofftargetHunter.jl --help</code></pre><h2 id="Example"><a class="docs-heading-anchor" href="#Example">Example</a><a id="Example-1"></a><a class="docs-heading-anchor-permalink" href="#Example" title="Permalink"></a></h2><p>TODO</p><h2 id="Public-Interface"><a class="docs-heading-anchor" href="#Public-Interface">Public Interface</a><a id="Public-Interface-1"></a><a class="docs-heading-anchor-permalink" href="#Public-Interface" title="Permalink"></a></h2><p>You can also use CRISPRofftargetHunter as a normal julia package with the exported functions.</p><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.Motif" href="#CRISPRofftargetHunter.Motif"><code>CRISPRofftargetHunter.Motif</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>Motif(     alias::String,     fwdmotif::String,      fwdpam::String,      forward_strand::Bool = true,      reverse_strand::Bool = true,      distance::Int = 4,      extends5::Bool = true)</code></p><p><code>Motif(alias::String)</code></p><p>Motif defines what we search on the genome, what can be identified as an off-target.</p><p><strong>Arguments</strong></p><p><code>alias</code> - alias of the motif for easier identification e.g. Cas9</p><p><code>fwdmotif</code> - Motif that indicates where is PAM inside <code>fwdpam</code>.     For example for Cas9 it is 20*N + XXX: NNNNNNNNNNNNNNNNNNNNXXX</p><p><code>fwdpam</code>   - Motif in 5&#39;-3&#39; that will be matched on the reference (without the X).              For example for Cas9 it is 20*X + NGG:              XXXXXXXXXXXXXXXXXXXXNGG</p><p><code>forward</code>  - If false will not match to the forward reference strand.</p><p><code>reverse</code>  - If false will not match to the reverse reference strand.</p><p><code>distance</code> - How many extra nucleotides are needed for a search? This              will indicate within what distance we can search for off-targets.              When we don&#39;t have those bases we use DNA_Gap.</p><p><code>extend5</code>  - Defines how off-targets will be aligned to the guides and where              extra nucleotides will be added for alignment within distance. Whether              to extend in the 5&#39; and 3&#39; direction. Cas9 is extend5 = true.</p><pre><code class="nohighlight">Example for Cas9 where we want to search for off-targets within distance of 4:
  alias:    Cas9
  fwdmotif: NNNNNNNNNNNNNNNNXXX
  fwdpam:   XXXXXXXXXXXXXXXXNGG
  forward:  true
  reverse:  true
  distance: 4
  extend5:  true</code></pre><p>Alignments will be performed from opposite to the extension direction (which is deifned by extend5).</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">Motif(&#39;Cas9&#39;)
Motif(&#39;Cas9&#39;, &#39;NNNNNNNNNNNNNNNNXXX&#39;, &#39;XXXXXXXXXXXXXXXXNGG&#39;. true, true, 3, true)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/motif.jl#L1-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.build_linearDB" href="#CRISPRofftargetHunter.build_linearDB"><code>CRISPRofftargetHunter.build_linearDB</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>build_linearDB(     name::String,     genomepath::String,     motif::Motif,     storagedir::String,     prefix_len::Int = 7)</code></p><p>Build a DB of offtargets for the given <code>motif</code>, DB groups off-targets by their prefixes.</p><p>Will return a path to the database location, same as <code>storagedir</code>. When this database is used for the guide off-target scan it is similar  to linear in performance, hence the name.</p><p>There is an optimization that if the alignment becomes imposible against the prefix we don&#39;t search the off-targets grouped inside the prefix. Therefore it is advantageous to select larger prefix than maximum  search distance, however in that case number of files also grows.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/db_linear.jl#L31-L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.search_linearDB" href="#CRISPRofftargetHunter.search_linearDB"><code>CRISPRofftargetHunter.search_linearDB</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>search_linearDB(storagedir::String, guides::Vector{LongDNA{4}}, dist::Int = 4; detail::String = &quot;&quot;)</code></p><p>Will search the previously build database for the off-targets of the <code>guides</code>.  Assumes your guides do not contain PAM, and are all in the same direction as  you would order from the lab e.g.:</p><p>` 5&#39; - ...ACGTCATCG NGG - 3&#39;  -&gt; will be input: ...ACGTCATCG      guide        PAM</p><p>3&#39; - CCN GGGCATGCT... - 5&#39;  -&gt; will be input: ...AGCATGCCC      PAM guide `</p><p><strong>Arguments</strong></p><p><code>dist</code> - defines maximum levenshtein distance (insertions, deletions, mismatches) for which off-targets are considered.   <code>detail</code> - path and name for the output file. This search will create intermediate  files which will have same name as detail, but with a sequence prefix. Final file will contain all those intermediate files. Leave <code>detail</code> empty if you are only  interested in off-target counts returned by the linearDB.  </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/db_linear.jl#L162-L185">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.build_sketchDB" href="#CRISPRofftargetHunter.build_sketchDB"><code>CRISPRofftargetHunter.build_sketchDB</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>build_sketchDB(     name::String,      genomepath::String,      motif::Motif,     storagedir::String,     probability_of_error::Float64 = 0.001;      max_count::Int = 255)</code></p><p>Build Count-Min-Sketch database that can never under-estimate the counts of off-targets.</p><p><strong>Arguments</strong></p><p><code>name</code> - Your name for this database.   <code>genomepath</code> - Path to the genome in either .2bit or .fa format. <code>motif</code>- See Motif type.   <code>storagedir</code>- Where should be the output folder.   <code>probability_of_error</code>- This is error level for each individual off-target search.      This error will be propagated when doing searches for larger distances and therefore      will not be perfectly reflected in the results.  </p><p><code>max_count</code> - What is the maximum count for each specific off-target in the genome. We     normally care only for guides that have low off-target count, therefore we can keep this     value also low, this saves space and memory usage.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/db_sketch.jl#L49-L74">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.search_sketchDB" href="#CRISPRofftargetHunter.search_sketchDB"><code>CRISPRofftargetHunter.search_sketchDB</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>search_sketchDB(     storagedir::String,     guides::Vector{LongDNA{4}},     dist::Int = 2)</code></p><p>For each of the input guides will check in the sketchDB  what value Count-Min Sketch has for it in the first, and  in the second column will output kmer sum. Results  are estimations of offtarget counts in the genome.</p><p>If CMS column is 0, it is guaranteed this guide has  no 0-distance off-targets in the genome!</p><p>Also, maximum count for each off-target in the database is capped, to <code>max_count</code> specified during building of sketchDB. This means  that counts larger than <code>max_count</code> (default 255) are no longer estimating correclty, but we know at least 255 off-targets exist in  the genome. Its likely you would not care for those guides anyhow.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/db_sketch.jl#L109-L128">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.build_treeDB" href="#CRISPRofftargetHunter.build_treeDB"><code>CRISPRofftargetHunter.build_treeDB</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>build_treeDB(     name::String,     genomepath::String,     motif::Motif,     storagedir::String,     prefix_len::Int = 7)</code>  </p><p>Build a Vantage Point tree DB of offtargets for the given <code>motif</code>, DB groups off-targets by their prefixes, each prefix has its own Vantage Point tree.</p><p>Will return a path to the database location, same as <code>storagedir</code>.</p><p>There is an optimization that if the alignment becomes imposible against the prefix we don&#39;t search the off-targets grouped inside the prefix. Therefore it is advantageous to select larger prefix than maximum  search distance, however in that case number of files also grows.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/db_tree.jl#L94-L113">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.search_treeDB" href="#CRISPRofftargetHunter.search_treeDB"><code>CRISPRofftargetHunter.search_treeDB</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>search_treeDB(storagedir::String, guides::Vector{LongDNA{4}}, dist::Int = 4; detail::String = &quot;&quot;)</code></p><p>Will search the previously build database for the off-targets of the <code>guides</code>.  Assumes your guides do not contain PAM, and are all in the same direction as  you would order from the lab e.g.:</p><pre><code class="nohighlight">5&#39; - ...ACGTCATCG NGG - 3&#39;  -&gt; will be input: ...ACGTCATCG
     guide        PAM

3&#39; - CCN GGGCATGCT... - 5&#39;  -&gt; will be input: ...AGCATGCCC
     PAM guide</code></pre><p><strong>Arguments</strong></p><p><code>dist</code> - defines maximum levenshtein distance (insertions, deletions, mismatches)  for which off-targets are considered.</p><p><code>detail</code> - path and name for the output file. This search will create intermediate  files which will have same name as detail, but with a sequence prefix. Final file  will contain all those intermediate files. Leave <code>detail</code> empty if you are only  interested in off-target counts returned by the searchDB.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/db_tree.jl#L237-L261">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CRISPRofftargetHunter.inspect_treeDB" href="#CRISPRofftargetHunter.inspect_treeDB"><code>CRISPRofftargetHunter.inspect_treeDB</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>inspect_treeDB(storagedir::String; levels::Int = 5, inspect_prefix::String = &quot;&quot;)</code></p><p>See small part of the full vantage point tree of the treeDB.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JokingHero/CRISPRofftargetHunter.jl/blob/619db9fb95743ad0acaa2b0c719e7e2ae2572cb6/src/db_tree.jl#L379-L383">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Saturday 19 June 2021 08:29">Saturday 19 June 2021</span>. Using Julia version 1.6.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
