<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Utils · ARTEMIS.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/theme.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img class="docs-light-only" src="../assets/logo.png" alt="ARTEMIS.jl logo"/><img class="docs-dark-only" src="../assets/logo-dark.png" alt="ARTEMIS.jl logo"/></a><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">General</a></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../abstract_gRNA/">Abstract gRNA</a></li><li><a class="tocitem" href="../find_potential_ot/">Find potential off-targets</a></li><li><a class="tocitem" href="../align_gRNA/">Align gRNA and off-target</a></li><li><a class="tocitem" href="../alignment_free/">Alignment-free filters for gRNAs</a></li><li><a class="tocitem" href="../find_ot/">Find all off-targets</a></li><li class="is-active"><a class="tocitem" href>Utils</a><ul class="internal"><li><a class="tocitem" href="#gRNAs-and-kmers"><span>gRNAs and kmers</span></a></li><li><a class="tocitem" href="#Persistence"><span>Persistence</span></a></li><li><a class="tocitem" href="#Summarize-off-targets"><span>Summarize off-targets</span></a></li><li><a class="tocitem" href="#Proximity-filter"><span>Proximity filter</span></a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>Utils</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Utils</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Utils"><a class="docs-heading-anchor" href="#Utils">Utils</a><a id="Utils-1"></a><a class="docs-heading-anchor-permalink" href="#Utils" title="Permalink"></a></h1><h2 id="gRNAs-and-kmers"><a class="docs-heading-anchor" href="#gRNAs-and-kmers">gRNAs and kmers</a><a id="gRNAs-and-kmers-1"></a><a class="docs-heading-anchor-permalink" href="#gRNAs-and-kmers" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.getseq" href="#ARTEMIS.getseq"><code>ARTEMIS.getseq</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>getseq(n = 20, letters = [&#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;])</code></p><p>Randomize sequence of length <code>n</code> from <code>letters</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.as_kmers" href="#ARTEMIS.as_kmers"><code>ARTEMIS.as_kmers</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>as_kmers(x::LongDNA{4}, kmer_size::Int)</code></p><p>Transforms <code>x</code> into vector of kmers of size <code>kmer_size</code>.  All ambiguous bases will be expanded.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; as_kmers(dna&quot;ACTGG&quot;, 4)
2-element Vector{LongSequence{DNAAlphabet{4}}}:
 ACTG
 CTGG</code></pre></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.as_skipkmers" href="#ARTEMIS.as_skipkmers"><code>ARTEMIS.as_skipkmers</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>as_skipkmers(x::LongDNA{4}, kmer_size::Int)</code></p><p>Transforms <code>x</code> into vector of skip-kmers of size <code>kmer_size</code>.  All ambiguous bases will be expanded.  Leftover-bases are ignored!</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; as_skipkmers(dna&quot;ACTGG&quot;, 2)
2-element Vector{LongSequence{DNAAlphabet{4}}}:
 AC
 TG</code></pre></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.all_kmers" href="#ARTEMIS.all_kmers"><code>ARTEMIS.all_kmers</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>all_kmers(size = 4; alphabet = [DNA_A, DNA_C, DNA_G, DNA_T]</code></p><p>Make a list of all possible kmers with given<code>size</code> using bases in the <code>alphabet</code>. </p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; all_kmers(2; alphabet = [DNA_A, DNA_N])
4-element Vector{LongSequence{DNAAlphabet{4}}}:
 AA
 AN
 NA
 NN</code></pre></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.minkmersize" href="#ARTEMIS.minkmersize"><code>ARTEMIS.minkmersize</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>minkmersize(len::Int = 20, d::Int = 4)</code></p><p>Pigeon hole principle: minimum k-mer size that is required for two strings of size <code>len</code> to be aligned within distance of <code>d</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; minkmersize(20, 3)
5

julia&gt; minkmersize(20, 6)
2</code></pre></div></section></article><h2 id="Persistence"><a class="docs-heading-anchor" href="#Persistence">Persistence</a><a id="Persistence-1"></a><a class="docs-heading-anchor-permalink" href="#Persistence" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.save" href="#ARTEMIS.save"><code>ARTEMIS.save</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>save(object::Any, destination::String)</code></p><p>Uses julia serializer to save the data to binary format. Read more about <a href="https://docs.julialang.org/en/v1/stdlib/Serialization/">serialization</a>. Notice that:</p><ol><li>This function will overwrite <code>destination</code>! </li><li>This serialization is dependent on julia build! This means files can fail to work when reloaded across different julia builds.</li></ol></div></section></article><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.load" href="#ARTEMIS.load"><code>ARTEMIS.load</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>load(destination::String)</code></p><p>Load file saved with <code>save</code> function. This <strong>may not</strong> load properly files saved in other julia builds.</p></div></section></article><h2 id="Summarize-off-targets"><a class="docs-heading-anchor" href="#Summarize-off-targets">Summarize off-targets</a><a id="Summarize-off-targets-1"></a><a class="docs-heading-anchor-permalink" href="#Summarize-off-targets" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.summarize_offtargets" href="#ARTEMIS.summarize_offtargets"><code>ARTEMIS.summarize_offtargets</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_overlapping(res::DataFrame, distance::Int)</code></pre><p>Filter overlapping off-targets. Remember that off-targets have their start relative to the PAM location.</p><p><strong>Arguments</strong></p><p><code>res</code> - DataFrame created by one of the off-target finding methods, it contains columns      such as <code>:guide,  :chromosome, :strand, :distance, :start</code>. </p><p><code>distance</code> - To what distance from the <code>:start</code> do we consider the off-target to be overlapping?</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">using ARTEMIS, BioSequences

# make a temporary directory
tdir = tempname()
db_path = joinpath(tdir, &quot;linearDB&quot;)
mkpath(db_path)

# use ARTEMIS example genome
artemis_path = splitpath(dirname(pathof(ARTEMIS)))[1:end-1]
genome = joinpath(vcat(artemis_path, 
    &quot;test&quot;, &quot;sample_data&quot;, &quot;genome&quot;, &quot;semirandom.fa&quot;))

# build a linearDB
build_linearDB(
    &quot;samirandom&quot;, genome, 
    Motif(&quot;Cas9&quot;), 
    db_path)

# load up example gRNAs
guides_s = Set(readlines(joinpath(vcat(artemis_path, 
    &quot;test&quot;, &quot;sample_data&quot;, &quot;crispritz_results&quot;, &quot;guides.txt&quot;))))
guides = LongDNA{4}.(guides_s)
    
# finally, make results!
res_path = joinpath(tdir, &quot;linearDB&quot;, &quot;results.csv&quot;)
search_linearDB(db_path, guides, res_path; distance = 3)

# load results
using DataFrames, CSV
res = DataFrame(CSV.File(res_path))

# filter results by close proximity
res = filter_overlapping(res, 23)

# summarize results into a table of counts by distance
summary = summarize_offtargets(res, 3)
</code></pre></div></section></article><h2 id="Proximity-filter"><a class="docs-heading-anchor" href="#Proximity-filter">Proximity filter</a><a id="Proximity-filter-1"></a><a class="docs-heading-anchor-permalink" href="#Proximity-filter" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ARTEMIS.filter_overlapping" href="#ARTEMIS.filter_overlapping"><code>ARTEMIS.filter_overlapping</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_overlapping(res::DataFrame, distance::Int)</code></pre><p>Filter overlapping off-targets. Remember that off-targets have their start relative to the PAM location.</p><p><strong>Arguments</strong></p><p><code>res</code> - DataFrame created by one of the off-target finding methods, it contains columns      such as <code>:guide,  :chromosome, :strand, :distance, :start</code>. </p><p><code>distance</code> - To what distance from the <code>:start</code> do we consider the off-target to be overlapping?</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">using ARTEMIS, BioSequences

# make a temporary directory
tdir = tempname()
db_path = joinpath(tdir, &quot;linearDB&quot;)
mkpath(db_path)

# use ARTEMIS example genome
artemis_path = splitpath(dirname(pathof(ARTEMIS)))[1:end-1]
genome = joinpath(vcat(artemis_path, 
    &quot;test&quot;, &quot;sample_data&quot;, &quot;genome&quot;, &quot;semirandom.fa&quot;))

# build a linearDB
build_linearDB(
    &quot;samirandom&quot;, genome, 
    Motif(&quot;Cas9&quot;), 
    db_path)

# load up example gRNAs
guides_s = Set(readlines(joinpath(vcat(artemis_path, 
    &quot;test&quot;, &quot;sample_data&quot;, &quot;crispritz_results&quot;, &quot;guides.txt&quot;))))
guides = LongDNA{4}.(guides_s)
    
# finally, make results!
res_path = joinpath(tdir, &quot;linearDB&quot;, &quot;results.csv&quot;)
search_linearDB(db_path, guides, res_path; distance = 3)

# load results
using DataFrames, CSV
res = DataFrame(CSV.File(res_path))

# filter results by close proximity
res = filter_overlapping(res, 23)

# summarize results into a table of counts by distance
summary = summarize_offtargets(res, 3)
</code></pre></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../find_ot/">« Find all off-targets</a><div class="flexbox-break"></div><p class="footer-message">Made with ♥ by Kornel Labun.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Tuesday 25 April 2023 21:51">Tuesday 25 April 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
